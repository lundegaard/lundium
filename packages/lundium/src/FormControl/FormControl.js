import React, { createRef, useContext, useEffect } from 'react';
import PropTypes from 'prop-types';
import { cx, flipIncludes, isNotNil, noop } from 'ramda-extension';
import invariant from 'invariant';

import FormGroupContext from '../contexts/FormGroupContext';
import isFilled from '../utils/isFilled';

const isTypeCheckable = flipIncludes(['radio', 'checkbox']);

const FormControl = ({
	component: Component,
	inputRef = createRef(),
	onBlur = noop,
	onFocus = noop,
	checked,
	children,
	className,
	name,
	size,
	type,
	value,
	...rest
}) => {
	const {
		onFilled,
		onEmpty,
		hasError,
		hasFloatingLabel,
		onFocus: onFocusContext,
		onBlur: onBlurContext,
	} = useContext(FormGroupContext);

	useEffect(() => {
		if (isFilled(value)) {
			onFilled();
		} else {
			onEmpty();
		}
	}, [onEmpty, onFilled, value]);

	const handleFocus = e => {
		onFocusContext();
		return onFocus(e);
	};

	const handleBlur = e => {
		onBlurContext();
		return onBlur(e);
	};

	const isCheckable = isTypeCheckable || isNotNil(checked);

	invariant(
		isCheckable || !hasFloatingLabel || value !== undefined,
		`FormControl with name "${name}" must be a controlled component. ` +
			'This is necessary for floating labels to function properly.',
	);

	return (
		<Component
			checked={checked}
			className={cx(className, 'form-control', size && `form-control-${size}`, {
				'is-invalid': hasError,
			})}
			name={name}
			type={type}
			value={value}
			ref={inputRef}
			{...rest}
			onBlur={handleBlur}
			onFocus={handleFocus}
		>
			{children}
		</Component>
	);
};

FormControl.propTypes = {
	/** Checked attribute of the form control. */
	checked: PropTypes.bool,
	/** Children to be rendered in the main container. */
	children: PropTypes.node,
	/** Class name for root. */
	className: PropTypes.string,
	/** Component used as form control. */
	component: PropTypes.elementType.isRequired,
	/** This prop can be used to pass a ref callback to the input element. */
	inputRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
	/** Name of the form control. */
	name: PropTypes.string.isRequired,
	/** Function to call when the form control emits a `blur` event. */
	onBlur: PropTypes.func,
	/** Function to call when the form control emits a `focus` event. */
	onFocus: PropTypes.func,
	/** Size of the form control. */
	size: PropTypes.oneOf(['sm', 'lg']),
	/** Type attribute of the form control. */
	type: PropTypes.string,
	/** Value of the form control. */
	value: PropTypes.any,
};

export default FormControl;
