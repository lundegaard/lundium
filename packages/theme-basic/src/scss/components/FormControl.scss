@import '../core';

/*
 -  Form control
 -  (<input class="form-control" type="*" />)
 */


.form-control {
	&:not(:disabled):not(:read-only):hover {
		border-color: $input-hover-border-color;
	}

	&[disabled],
	&[readonly],
	&[disabled] + .control-label,
	&[readonly] + .control-label {
		color: $control-label-disabled-color;
	}

	&.is-invalid,
	&:invalid {
		color: $input-invalid-color;
		background: $input-bg;
		padding-right: $input-padding-x;
	}

	&.form-control--editable {
		padding: logical 0;
		height: auto;
	}

	&.empty-checkbox + .control-label {
		text-indent: -9999px;
		width: 0;
	}


	&__legend {
		position: absolute;
		font-size: 1.2rem;
		inset-inline-end: ( $spacer * .75 );
		transform: translateY(4px);
		color: $info;
	}

	&::-ms-clear,
	&::-ms-reveal {
		display: none;
	}
}

