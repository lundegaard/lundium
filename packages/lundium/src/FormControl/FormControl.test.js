import React from 'react';
import { shallow } from 'enzyme';

import FormControl from './FormControl';

describe('<FormControl />', () => {
	const requiredProps = {
		component: 'input',
		name: 'formControlField',
	};

	it('shallow', () => {
		const wrapper = shallow(<FormControl {...requiredProps} />);
		expect(wrapper).toMatchSnapshot();
	});
});
