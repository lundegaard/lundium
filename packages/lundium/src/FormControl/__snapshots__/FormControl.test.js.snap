// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormControl /> shallow 1`] = `
<input
  className="form-control"
  name="formControlField"
  onBlur={[Function]}
  onFocus={[Function]}
/>
`;
